### ISSUE TITLE
Deluge plugins

### ISSUE THREAD

**Issue by MagnaCapax** [MagnaCapax/PMSS#23]
Time to check them and include the pertinent ones.

**Comment by MagnaCapax** [comment]
## Adversarial Review

**Issue:** #23 — Deluge plugins
**Reviewed:** 2026-04-16
**Verdict:** NEEDS-INVESTIGATION

> Prior reviews: 2026-04-01 (NEEDS-INVESTIGATION), 2026-04-09 (NEEDS-INVESTIGATION).
> This review covers the material change since 2026-04-09.

---

### Material Change Since Last Review

**Issue #296 (Enforce managed configuration for qBittorrent and Deluge) was closed on 2026-04-10. The Deluge managed-config enforcement — the blocking prerequisite identified in both prior reviews — was silently dropped, not implemented.**

Reconstruction of what happened:

| Date | Event |
|------|-------|
| 2026-04-08 | Partial verification report explicitly lists: "Deluge managed config enforcement \| UNCLEAR \| — \| Not found in commit history." `complete-verify` label retained *pending full implementation*. |
| 2026-04-09 | No Deluge commits in git log. |
| 2026-04-10 | Final verification runs — covers qBittorrent only. Deluge acceptance criterion absent from the verification table entirely. Issue is unlocked and closed as PASS. |

The final verification comment for #296 contains no mention of Deluge, no `pmssDelugeApplyManagedConfig()` equivalent, no test assertions for Deluge managed-config. The issue title is "Enforce managed configuration for **qBittorrent and Deluge**." Only one of the two named subsystems was implemented and verified.

This is a **FALSE_COMPLETION_CLAIM** on the prerequisite. The closure of #296 has not resolved the blocking condition — it has made it invisible by removing it from the open issue tracker.

---

### Current Source State (Confirmed 2026-04-16)

Unchanged from both prior reviews. Zero Deluge changes in git log since 2026-04-09.

- `etc/seedbox/config/template.deluge.core.conf:18` — `"enabled_plugins": []`
- `etc/seedbox/config/template.deluge.core.nocache.conf:16` — `"enabled_plugins": []`
- `etc/seedbox/config/template.deluge.web.conf:8` — `"enabled_plugins": []`
- `scripts/lib/user/deluge.php` — `userConfigureDeluge()` unconditionally writes `core.conf` from template (`file_put_contents(...)` at line 117); no managed-key preservation logic exists
- `scripts/lib/update/apps/deluge.php` — zero plugin references
- No function analogous to `pmssQbittorrentApplyManagedConfig()` exists for Deluge anywhere in the codebase

The mechanism that would overwrite any future plugin configuration on every maintenance cycle remains fully active. Implementing #23 today would result in plugins being silently reset to `[]` on the next `userConfigureDeluge()` call.

---

### Adversarial Findings

**[C] The prerequisite blocking condition is now untracked, not resolved**

Before April 10, #296 was open with an explicit "Deluge coverage pending" note. The Deluge gap had a home in the issue tracker. After April 10, it does not. The infrastructure gap is real and unchanged; only its visibility has changed. If #23 is implemented without first rebuilding the Deluge managed-config infrastructure, the result is deterministic: `enabled_plugins` will be zeroed on every maintenance cycle.

**[C] FALSE_COMPLETION_CLAIM on #296 creates a compounding risk for any downstream issue that listed #296 as a prerequisite**

Any other open issue that referenced #296's Deluge coverage as a dependency now incorrectly appears unblocked. #23 is the most direct example. An implementer reading the issue tracker today would see #296 as closed/PASS and proceed without understanding that the Deluge half was never delivered.

**[M] The security context has not changed — and remains the primary gate**

`docs/incidents/2025-12-15-deluge-system-services-debian-deluged.md` still records the plugin/command-path exploit chain. The attack surface analysis from prior reviews stands. No security assessment of specific plugin candidates exists. No per-plugin exclusion list exists. These gaps are unchanged and are independent of the #296 prerequisite.

---

### Caller Analysis

No new callers since 2026-04-09. Full enumeration from prior review holds:
- `scripts/util/userConfig.php:167` is the sole non-test caller of `userConfigureDeluge()`
- Zero plugin management call sites exist anywhere in the codebase

---

### Recommendations

1. **File a dedicated issue for Deluge managed-config enforcement.** The Deluge half of #296 needs its own tracking issue now that #296 is closed. Without this, the gap has no home and will remain unimplemented indefinitely. This is a prerequisite for #23 and should be explicitly linked.

2. **#23 remains blocked** until: (a) Deluge managed-config enforcement is implemented (new issue needed), (b) a named plugin candidate list with per-plugin security assessment exists, (c) an ADR specifies the architectural home for plugin management.

3. The deletion alternative from the 2026-04-09 review remains valid: document that PMSS intentionally defers plugin management to users, add a comment to the templates explaining the security rationale, close this issue. This resolves the question without creating new attack surface.

---

### Worth-Implementing Analysis

- **Cost**: High — unchanged from prior reviews
- **Benefit**: Undefined — unchanged from prior reviews; no user demand linked in two years
- **Prerequisite gap**: The managed-config foundation (the #296 Deluge half) is unimplemented and now untracked
- **Verdict rationale**: NEEDS-INVESTIGATION maintained. The issue cannot proceed until (a) a new issue covers the #296 Deluge gap and reaches `complete-verify`, and (b) a concrete plugin candidate list with security assessment is written.

---

### Anti-Pattern Hunt

| Class | Verdict | Evidence |
|---|---|---|
| `REGEX_FOR_SEMANTICS` | PASS | Not applicable |
| `PREMATURE_OPTIMIZATION` | FAIL | Extending plugin infrastructure that does not exist, zero measured user demand |
| `FALSE_COMPLETION_CLAIM` | **FAIL on prerequisite #296** | #296 closed with qBittorrent-only verification; Deluge criterion present in April 8 partial report, absent from April 10 final report |
| `OPERATOR_HANDOFF_ANTIPATTERN` | PASS | Not present in this issue |
| `FABRICATED_AUTHORITY_OR_PATH` | PASS | All cited paths verified against live source |
| `FAIL_OPEN_SAFETY_BEHAVIOR` | **FAIL risk if naively implemented** | Plugin deployment without exclusion list defaults to fail-open on execution-capable plugins |
| `DRY_RUN_THEATER` | PASS | Not applicable |
| `INVESTIGATION_LOOP_THEATER` | PASS | New information present since last review |

---

### Decomposition Targets

No new decomposition targets since 2026-04-09. Prior review targets in `scripts/lib/user/deluge.php:87-162` and `scripts/lib/update/apps/deluge.php` remain valid.

---

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review

**Issue:** #23 — Deluge plugins
**Reviewed:** 2026-04-29
**Verdict:** REWORK

> Prior reviews: 2026-04-01 (NEEDS-INVESTIGATION), 2026-04-09 (NEEDS-INVESTIGATION), 2026-04-16 (NEEDS-INVESTIGATION).
> This review covers material changes since 2026-04-16.

---

### Material Changes Since 2026-04-16

**Issue #462 (Deluge managed-config enforcement) is now closed as complete-verify (2026-04-19).**

Commit `693df1a9` introduces `scripts/lib/user/delugeManagedConfig.php` (170 lines), implementing `pmssDelugeApplyManagedConfig()`. The function surgically patches four PMSS-managed keys in the live `core.conf` (`max_active_downloading`, `max_active_limit`, `max_connections_global`, `max_upload_slots_global`) while preserving all other keys — including `enabled_plugins`. The `pmssDelugeManagedConfigEntries()` whitelist deliberately excludes `enabled_plugins`, meaning any future plugin deployment would survive the maintenance cycle intact.

This resolves the primary infrastructure blocker identified in all three prior reviews.

**Architecture clarification: `userConfigureDeluge()` is provisioning-only, not a maintenance-cycle concern.**

Prior reviews framed the risk as "the maintenance cycle zeros `enabled_plugins` on every run." This framing was imprecise. `userConfigureDeluge()` (which unconditionally writes the full template at `scripts/lib/user/deluge.php:117`, including `"enabled_plugins": []`) is invoked only during provisioning — through `scripts/lib/user/add/userConfigApply.php` and `scripts/recreateUser.php` — not from the ongoing maintenance cycle (`update-step2.php` → `scripts/lib/update/users/http.php`). The maintenance cycle calls only `pmssDelugeApplyManagedConfig()`, which now preserves `enabled_plugins`. Plugins deployed at provisioning time will survive routine maintenance passes.

**Issue #474 filed 2026-04-28: Deluge hostlist token drift (customer-reported).**

`hostlist.conf` auth token drifts from the daemon's `auth` file when `auth` is regenerated after initial provisioning, causing the web UI to silently show an empty torrent grid. This is a separate Deluge auth infrastructure bug — not a blocker for #23 — but is evidence of ongoing Deluge config fragility. It is tracked independently and references #462 for the broader managed-config context.

---

### Adversarial Findings

**[M] Three rounds of NEEDS-INVESTIGATION completed — the remaining gap is a spec problem, not an investigation problem**

The prior three reviews established: (a) zero plugin management infrastructure exists, (b) the managed-config prerequisite (#462) was needed to prevent plugins being zeroed on maintenance cycles, (c) no plugin candidates had been named. Issue #462 resolves (b). Items (a) and (c) are spec-time decisions, not research outcomes — they require the operator to name specific plugins and conduct per-plugin security assessments. After 28 months, this issue still contains no plugin names, no security assessment, and no deployment design. A fourth NEEDS-INVESTIGATION verdict would be investigation-loop theater; the correct escalation is REWORK.

**[C] Security context unchanged — confirmed exploit class in incident record, no exclusion list written**

`docs/incidents/2025-12-15-deluge-system-services-debian-deluged.md` records a plugin or command path as the confirmed attack class in the December 2025 PM infrastructure compromise. No per-plugin security assessment exists in this issue, in any ADR, or in the codebase. The Execute plugin — Deluge's canonical local-command-execution plugin — remains unaddressed by this issue. Any implementation that does not explicitly enumerate and exclude execution-capable plugins risks expanding exactly the attack surface documented in that incident report.

**[M] Version compatibility matrix still absent**

`pmssDelugeCoreTemplatePath()` at `scripts/lib/user/deluge.php:22` already branches on `$distro['version'] >= 12` because libtorrent 2.0 (Debian 12+) changed the plugin API surface. Debian 10 ships Deluge 2.0.5 (pip-built); Debian 11–12 ship Deluge 2.1.1 (apt). Plugin `.egg` compatibility and available plugin versions differ between these installations. No plugin list specifying per-version compatibility exists.

**[L] `pmssDelugeConfigMutate()` is silently no-op on missing or unreadable config**

The new `pmssDelugeConfigMutate()` at `scripts/lib/user/delugeManagedConfig.php:120` returns `false` when `core.conf` is absent, a symlink, or unparseable. For the current limits-enforcement use case this is correct — provisioning writes the file first. For any future plugin-enabling implementation, a missing config would silently skip plugin deployment with no error signal. This behavior requires explicit handling in any plugin deployment design.

---

### Big Questions

**Worst case if implemented without a proper spec:** An execution-capable plugin is included in the provisioned default list without per-plugin security review. Every new Deluge account receives this plugin at provisioning. An attacker with web UI access finds the Execute plugin pre-installed and uses it. The December 2025 incident documents exactly this attack class against PM infrastructure, and enabling execution-capable plugins fleet-wide would directly extend that surface. Blast radius: all Deluge-enabled accounts.

**Inversion:** To guarantee maximum damage: leave the issue body as five words, let the Build agent select plugins by community popularity heuristics, skip security review, deploy fleet-wide at provisioning. The current issue body creates precisely the conditions for this outcome.

**Regret test (6 months):** A plugin shipped as a PMSS default develops an upstream vulnerability — or was already abandoned upstream — and is deployed to every Deluge user. Removing it from existing accounts requires a coordinated targeted `core.conf` mutation pass fleet-wide or full reprovisioning, both operationally non-trivial with user impact.

**Removal inversion:** If PMSS already shipped default Deluge plugins and someone proposed removing them, users who had configured workflows around those plugins would object. The removal cost is non-trivial once deployed. The cost of not shipping is `enabled_plugins: []` — the current state, which has produced zero documented user complaints in 28 months. The asymmetry favors caution.

**Requirements challenge:** The dumbest implicit requirement is "PMSS should own the Deluge plugin surface at all." Deluge's Web UI allows users to install and manage plugins themselves. This issue proposes centralizing management of a capability users already exercise individually — with no evidence they are failing to do so, and no user demand signal attached.

**Deletion opportunity:** Remove the `plugins_location` token from all three Deluge templates and add a comment explaining that PMSS intentionally leaves plugin management to users, citing the December 2025 security context. A 3-line change that reduces attack surface, documents intent, and closes this issue honestly. The alternative to implementing this feature is actively declining to own the fleet plugin attack surface — which may be the correct decision.

---

### Caller Analysis

No existing plugin management code to enumerate. This is entirely new functionality.

Current confirmed source state (2026-04-29):
- `etc/seedbox/config/template.deluge.core.conf:18` — `"enabled_plugins": []`
- `etc/seedbox/config/template.deluge.core.nocache.conf:16` — `"enabled_plugins": []`
- `etc/seedbox/config/template.deluge.web.conf:8` — `"enabled_plugins": []`
- `scripts/lib/user/delugeManagedConfig.php` — new since 2026-04-16; `enabled_plugins` intentionally absent from managed entries; maintenance cycle preserves it
- `scripts/lib/user/deluge.php:117` — `userConfigureDeluge()` unconditionally writes full template; provisioning-only call site
- `scripts/lib/update/users/http.php:36` — maintenance cycle calls `pmssDelugeApplyManagedConfig()`; does not touch `enabled_plugins`

---

### Worth-Implementing Analysis

- **Cost**: High. No plugin management scaffold exists. Implementation requires: named plugin list, per-plugin security assessment, version compatibility matrix (Debian 10/11/12), HTTPS + SHA256 integrity verification for each `.egg`, deployment architecture ADR, and integration test coverage. The reprovisioning interaction (`recreateUser.php` zeros and rewrites `core.conf`) also needs explicit handling.
- **Benefit**: Undefined. No linked support tickets. No named plugins. Zero documented user complaints about missing Deluge plugins in 28 months.
- **Frequency of problem**: Cannot be assessed — no user demand signal exists.
- **Alternative**: Document intentional user-managed plugin policy. ~2 hours, zero attack surface impact, honest closure.
- **Verdict rationale**: Cost/benefit ratio is unfavorable given undefined demand and serious security requirements. The infrastructure prerequisite (#462) is now resolved, removing the last technical blocker — but the spec has never been written. The issue needs a complete rewrite as an implementation specification before any build work begins. REWORK.

---

### Merit Check (Tier A)

**1. Concrete present demand.** No linked support tickets. No user has reported a missing Deluge plugin via any tracked channel in 28 months. Theoretical — no present demand.

**2. Function.** When shipped, enables a named set of Deluge plugins as PMSS fleet defaults, written into `enabled_plugins` in `core.conf` at provisioning time, with `.egg` artifacts placed in the user's plugins directory.

**3. Success metric.** Not definable without named plugins. No independent pre-existing signal exists to measure a delta against. This absence is itself a strong REWORK signal — nothing that cannot be measured can be said to have worked.

**4. Catastrophic if implemented (causal chain).** An execution-capable plugin reaches the default list without security vetting → deployed to all Deluge accounts at provisioning → any attacker with authenticated web UI access finds command execution pre-enabled → the attack class documented in `docs/incidents/2025-12-15-deluge-system-services-debian-deluged.md` applies directly. Blast radius: fleet-wide across all Deluge-enabled accounts.

**5. Catastrophic if NOT implemented (causal chain).** Users who want specific Deluge plugins install them via the Web UI, as they do today. Zero documented user harm from the current state. The catastrophic-if-NOT scenario is: no change from current state for another 28 months.

**6. Deletion alternative.** Remove `plugins_location` from all three core templates. Add a comment block explaining that PMSS intentionally defers plugin management to users, with a reference to the December 2025 incident. Lowest-cost resolution that honestly documents intent.

**7. Cost.** (a) Build: 3–5 engineering days including security research; no scaffold exists. (b) Maintenance: Each shipped plugin requires ongoing security monitoring; upstream CVE or project abandonment requires coordinated fleet response — minimum annual review per plugin. (c) Blast radius if broken: A provisioning regression affects all new Deluge accounts; a security regression affects existing accounts fleet-wide. (d) Undo cost: Removing an enabled plugin from existing accounts requires a targeted `core.conf` mutation pass fleet-wide or full reprovisioning — non-trivial rollback with user impact.

**8. Dependency blast (Hyrum).** If plugins are deployed at provisioning time, `recreateUser.php` calls `userConfigureDeluge()` which writes the full template (zeroing `enabled_plugins`), requiring reprovisioning to re-apply plugins as a separate step. Every caller of `userConfigureDeluge()` becomes a dependency of the plugin deployment logic.

**9. Requirements challenge (Musk Step 1).** The dumbest implicit requirement is "PMSS must centrally manage Deluge plugins." Deluge's Web UI allows users to install and manage plugins themselves. This issue proposes taking on fleet-wide ownership of a capability users already have, without evidence they want PMSS to own it on their behalf.

---

### Anti-Pattern Hunt

| Class | Verdict | Evidence |
|---|---|---|
| `REGEX_FOR_SEMANTICS` | PASS | Not applicable |
| `PREMATURE_OPTIMIZATION` | FAIL | Proposes fleet-wide plugin infrastructure with zero measured user demand; base feature need unproven |
| `FALSE_COMPLETION_CLAIM` | WARNING on #462 | QA PASS at 11:30 on 2026-04-19, followed by a separate runner SKIP at 18:57 the same day ("output too short — re-queued for E2E"). Issue is labeled `complete-verify` and the implementation is architecturally sound. The SKIP appears to be a stale re-queue artifact on an already-closed issue, not a code defect — but full E2E verification was not recorded as completed. |
| `OPERATOR_HANDOFF_ANTIPATTERN` | PASS | Not present |
| `FABRICATED_AUTHORITY_OR_PATH` | PASS | All cited file paths and line numbers verified against live source in this review session |
| `FAIL_OPEN_SAFETY_BEHAVIOR` | PARTIAL — note for implementation | `pmssDelugeConfigMutate()` is silent no-op on missing/unreadable config; acceptable for current limits enforcement but needs explicit handling in any plugin deployment design |
| `DRY_RUN_THEATER` | PASS | Not applicable |
| `INVESTIGATION_LOOP_THEATER` | PREVENTED | Verdict escalated to REWORK rather than issuing a fourth NEEDS-INVESTIGATION; infrastructure prerequisite now resolved; remaining gap is spec, not investigation |

---

### Recommendations

1. **Seriously consider the deletion path.** Add a comment to all three Deluge core templates explaining that PMSS intentionally defers plugin management to users, citing the security context. Close as won't-implement. In 28 months no user demand signal has appeared to justify the alternative path.

2. **If proceeding, replace the issue body with a proper implementation spec.** The current body provides no implementation basis. The rewrite must include: (a) a named list of candidate plugins, each assessed for execution capability, filesystem scope, network access, and upstream maintenance status; (b) explicit exclusion of the Execute plugin and any plugin enabling local command execution or arbitrary file writes outside the user's home; (c) a per-version compatibility matrix for Deluge 2.0.5 (Debian 10) and 2.1.1 (Debian 11/12); (d) an ADR specifying the deployment architecture — provisioning-time versus update-time, and how reprovisioning via `recreateUser.php` is handled; (e) HTTPS source URLs and SHA256 checksums for each plugin artifact.

3. **Issue #474 (hostlist token drift) is separately tracked and is not a blocker or dependency for this issue.**

---

### Decomposition Targets

No new targets beyond what prior reviews identified. Prior targets remain valid:
- **[REFACTOR]** `scripts/lib/user/deluge.php:81-165` (`userConfigureDeluge`) — 84-line function with seven distinct responsibilities. Split candidates: `pmssDelugeEnsureDirectories()`, `pmssDelugeWriteConfigs()`, `pmssDelugeRestartIfNeeded()`.
- **[REFACTOR]** `scripts/lib/update/apps/deluge.php` (408 lines) — inline executable section below `PMSS_DELUGE_NO_ENTRYPOINT` guard mixes orchestration with library-style helper functions. Splitting into a pure-library include and a thin executable would make each independently testable.

---

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Investigation Conclusion — #23 Deluge plugins

**Reviewed:** 2026-06-02 · **Status:** no plugin-management implementation exists · **Recommendation:** SHOULD-NOT-IMPLEMENT as written

The prior reviews on this thread converged. This pass re-verified their load-bearing points against current `HEAD` rather than trusting them — all hold, unchanged:

- `enabled_plugins` is `[]` in all three Deluge templates and has only ever been `[]` across the full git history; no plugin-management code exists in any Deluge file; the `plugins_location` directory is created but never populated.
- `docs/incidents/2025-12-15-deluge-system-services-debian-deluged.md` documents the plugin/command surface as a **real execution vector** in a confirmed compromise — not a theoretical risk.
- `AGENTS.md` bars `scripts/lib/update/apps/` installers from owning ongoing config, so there is no defined architectural home for plugin management.

**Genuinely new since the last review:** the managed-config prerequisite (#296 / #462) is now **closed**. `delugeManagedConfig.php` manages only the active/connection/upload-slot limits and `pre_allocate_storage` — it deliberately does **not** manage `enabled_plugins`, so a future plugin list would be *preserved* across maintenance cycles rather than silently zeroed. That removes the last *technical* blocker the early reviews raised — **but the recommendation is unchanged**, because the real blockers were never technical:

1. The issue body ("check them and include the pertinent ones") is spec-less: no named plugins, no per-plugin security assessment, no Debian 10/11/12 (Deluge 2.0.5 vs 2.1.1) compatibility matrix.
2. Shipping any execution-capable plugin (e.g. the Execute plugin) as a fleet default would extend exactly the surface used in the December 2025 incident.
3. No user demand in 28 months — Deluge's Web UI already lets users install plugins themselves.

**Recommendation — decline to own the fleet plugin attack surface** (YAGNI: real risk, no demand). The honest disposition is an operator product decision between:

- **(a) Close as won't-implement**, optionally adding a one-line template comment recording that PMSS intentionally leaves plugin management to users; or
- **(b) Rewrite this issue as a security-vetted implementation spec** — named candidates, per-plugin security review excluding execution-capable plugins, version-compatibility matrix, deployment ADR, and HTTPS+SHA256 artifact integrity.

No open issue supersedes this one, so it is left open for that operator decision rather than closed automatically.

— Väinämöinen

**Comment by MagnaCapax** [comment]
## Adversarial Review

**Issue:** #23 — Deluge plugins
**Reviewed:** 2026-06-07
**Verdict:** SHOULD-NOT-IMPLEMENT (as written) — reaffirmed; disposition label applied

This is a brief disposition update, not a re-derivation. The prior reviews on this thread converged; their load-bearing anchors were re-verified against current `HEAD` this pass and all hold unchanged:

- `enabled_plugins` is `[]` in all three Deluge templates and has only ever been `[]` across the full git history. No plugin-management code exists in any Deluge file; the `plugins_location` directory is created at provisioning but never populated.
- The managed-config layer (`scripts/lib/user/delugeManagedConfig.php`) governs only the active/connection/upload-slot limits and storage pre-allocation — it deliberately does **not** manage `enabled_plugins`. A future plugin list would therefore be preserved across maintenance cycles rather than silently zeroed. The last *technical* blocker is resolved.
- The plugin/command surface remains the confirmed execution vector in the documented December 2025 incident. The issue body is still a five-word memo: no named plugin candidates, no per-plugin security assessment, no Debian 10/11/12 (Deluge 2.0.5 vs 2.1.1) compatibility matrix.
- No Deluge commit since the last review touches the plugin surface — the only recent Deluge change hardens the config-write path through a symlink-safe writer and adds tests.

**New this pass:** the `should-not-implement` label has been applied to formalize the converged verdict. It reflects that the issue **as written** should not be built — it is **not** a closure. The honest disposition remains an operator product decision between:

- **(a)** close as won't-implement — optionally with a one-line template comment recording that PMSS intentionally leaves plugin management to users (zero added attack surface; no user demand in the issue's lifetime, and the Deluge Web UI already lets users install plugins themselves); or
- **(b)** rewrite this issue as a security-vetted implementation spec — named candidates, per-plugin security review explicitly excluding execution-capable plugins (e.g. the Execute plugin), a version-compatibility matrix, a deployment-architecture ADR, and HTTPS + SHA256 artifact integrity.

No open issue supersedes this one, so it is left open for that operator decision rather than closed automatically.

— Väinämöinen
